### PR TITLE
util/execdetails: refactor execdetails information of runtime collect (#18530)

### PR DESCRIPTION
**Pull request by ti-srebot**
cherry-pick #18530 to release-4.0

---

### What problem does this PR solve?

This PR refactor `execdetails` information of runtime collect. This will make it easy and efficient to add special runtime information for the separated executors.

### What is changed and how it works?

Before this PR, all executor runtime information was expressed by `execdetails.RuntimeStats`. But different executor maybe has different runtime information, then the `execdetails.RuntimeStats` becomes very huge and redundant.

This PR use `execdetails.RuntimeStats` interface instead of ``execdetails.RuntimeStats` struct.

### Related changes

- Need to cherry-pick to the release branch

### Check List 

Tests 

- No code, only refactor.

Side effects

- No

### Release note 

- No

### CODE REVIEW

**Comment by qw4990** [comment]
@crazycs520 Please fix the CI problem at your convenience.

File: executor/join.go [hunk 1]
```diff
@@ -880,3 +882,58 @@ func (e *NestedLoopApplyExec) Next(ctx context.Context, req *chunk.Chunk) (err e
 		}
 	}
 }
+
+// cacheInfo is used to save the concurrency information of the executor operator
+type cacheInfo struct {
```
Comment (SunRunAway) [inline executor/join.go:887]: Is it not used?

**Pull request merged**